### PR TITLE
Respect read-mode when using terminate-at-block during replay

### PR DESCRIPTION
**Pull request by jnordberg**
## Change Description

This adapts the changes from #10676 to also apply during replays.

## Change Type
**Select *ONE*:**
- [ ] Documentation

- [x] Stability bug fix

- [ ] Other

- [ ] Other - special case

## Testing Changes
**Select *ANY* that apply:**
- [ ] New Tests

- [ ] Existing Tests

- [ ] Test Framework

- [ ] CI System

- [ ] Other

## Consensus Changes
- [ ] Consensus Changes

## API Changes
- [ ] API Changes

## Documentation Additions
- [ ] Documentation Additions

### CODE REVIEW

**Comment by heifner** [comment]
Thanks for the contribution. Is there a benefit to this over just running in `read-mode=irreversible` and using `terminate-at-block` ? Which now works correctly with https://github.com/EOSIO/eos/pull/10676.

File: plugins/chain_plugin/chain_plugin.cpp [hunk 1]
```diff
@@ -1281,6 +1290,10 @@ void chain_plugin::plugin_initialize(const variables_map& options) {
       } );
 
       my->irreversible_block_connection = my->chain->irreversible_block.connect( [this]( const block_state_ptr& blk ) {
+         if (my->chain_config->stop_at_lib > 0 && blk->block_num >= my->chain_config->stop_at_lib) {
+            wlog( "Stopping at lib#${num}.", ("num", blk->block_num));
+            appbase::app().quit();
```
Comment (heifner) [inline plugins/chain_plugin/chain_plugin.cpp]: This simply signals application to shutdown. All queued posted jobs to the main thread complete which could potentially include some more blocks.

**Comment by jnordberg** [comment]
Does `--terminate-at-block` work in non replay mode now as well? I remember it throwing an error at me trying to use it without `--replay` which lead to having to add this patch, if that works this patch would be redundant yeah. Just need something to stop nodeos once seeing a specific lib regardless if it's replaying or not

**Comment by heifner** [comment]
> Does `--terminate-at-block` work in non replay mode now as well? I remember it throwing an error at me trying to use it without `--replay` which lead to having to add this patch, if that works this patch would be redundant yeah. Just need something to stop nodeos once seeing a specific lib regardless if it's replaying or not

At a quick glance it looks like it does not work for replay, but that should be a one line change to `replay_push_block` like the change to `push_block` of https://github.com/EOSIO/eos/pull/10676.

**Comment by jnordberg** [comment]
Like this?

**Comment by heifner** [comment]
> Like this?

Yes. Can you test that out and make sure it works as expected.

**Comment by jnordberg** [comment]
Seems to be working

Without patch using --replay-blockchain --read-mode=irreversible --terminate-at-block=50
```
info  2021-09-19T03:27:31.017 nodeos    controller.cpp:1895           replay_push_block    ] Reached configured maximum block 50; terminating
info  2021-09-19T03:27:31.017 nodeos    controller.cpp:448            replay               ] 49 irreversible blocks replayed
info  2021-09-19T03:27:31.017 nodeos    controller.cpp:452            replay               ] fork database head 50, root 50
info  2021-09-19T03:27:31.017 nodeos    controller.cpp:495            replay               ] replayed 49 blocks in 1 seconds, 23.63895833333333485 ms/block
```
Without patch using --replay-blockchain --read-mode=head --terminate-at-block=50
```
info  2021-09-19T03:29:26.559 nodeos    controller.cpp:1895           replay_push_block    ] Reached configured maximum block 50; terminating
info  2021-09-19T03:29:26.559 nodeos    controller.cpp:448            replay               ] 49 irreversible blocks replayed
info  2021-09-19T03:29:26.559 nodeos    controller.cpp:452            replay               ] fork database head 50, root 50
info  2021-09-19T03:29:26.559 nodeos    controller.cpp:495            replay               ] replayed 49 blocks in 1 seconds, 22.17833333333333101 ms/block
```

With patch using  --replay-blockchain --read-mode= irreversible --terminate-at-block=50
```
info  2021-09-19T03:36:42.850 nodeos    controller.cpp:1895           replay_push_block    ] Reached configured maximum block 50; terminating
info  2021-09-19T03:36:42.850 nodeos    controller.cpp:448            replay               ] 50 irreversible blocks replayed
info  2021-09-19T03:36:42.850 nodeos    controller.cpp:452            replay               ] fork database head 51, root 51
info  2021-09-19T03:36:42.851 nodeos    controller.cpp:495            replay               ] replayed 50 blocks in 1 seconds, 22.05844897959183726 ms/block
```

With patch using  --replay-blockchain --read-mode=head --terminate-at-block=50
```
info  2021-09-19T03:34:34.105 nodeos    controller.cpp:1895           replay_push_block    ] Reached configured maximum block 50; terminating
info  2021-09-19T03:34:34.105 nodeos    controller.cpp:448            replay               ] 50 irreversible blocks replayed
info  2021-09-19T03:34:34.105 nodeos    controller.cpp:452            replay               ] fork database head 50, root 50
info  2021-09-19T03:34:34.105 nodeos    controller.cpp:454            replay               ] resetting fork database with new last irreversible block as the new root: 00000033e5632dc2b9b432936de2bdf432ea8341a22ac3fb431bb03e6308f781
info  2021-09-19T03:34:34.106 nodeos    controller.cpp:495            replay               ] replayed 50 blocks in 1 seconds, 22.15940816326530438 ms/block
```

**Review by heifner — APPROVED** [review]
PR description needs to be updated.
Also needs to be applied to `release/2.1.x` and `release/2.2.x` branches.

**Comment by heifner** [comment]
Thanks again for this PR. I've created 2.1.x and 2.2.x versions of the fix.

**Pull request merged**